src/Http/Http.Results/src/ResultsOfTHelper.cs

```diff
@@ -1,21 +1,51 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
+using System.Runtime.CompilerServices;
 using Microsoft.AspNetCore.Builder;
 using Microsoft.AspNetCore.Http.Metadata;
 
 namespace Microsoft.AspNetCore.Http;
 
 internal static class ResultsOfTHelper
 {
+    public const DynamicallyAccessedMemberTypes RequireMethods = DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods;
     private static readonly MethodInfo PopulateMetadataMethod = typeof(ResultsOfTHelper).GetMethod(nameof(PopulateMetadata), BindingFlags.Static | BindingFlags.NonPublic)!;
 
-    public static void PopulateMetadataIfTargetIsIEndpointMetadataProvider<TTarget>(MethodInfo method, EndpointBuilder builder)
+    public static void PopulateMetadataIfTargetIsIEndpointMetadataProvider<[DynamicallyAccessedMembers(RequireMethods)] TTarget>(MethodInfo method, EndpointBuilder builder)
```

Comment:
No. It's called from ResultsOfT types to populate metadata.

For example:

```csharp
static void IEndpointMetadataProvider.PopulateMetadata(MethodInfo method, EndpointBuilder builder)
{
    ArgumentNullException.ThrowIfNull(method);
    ArgumentNullException.ThrowIfNull(builder);

    ResultsOfTHelper.PopulateMetadataIfTargetIsIEndpointMetadataProvider<TResult1>(method, builder);
    ResultsOfTHelper.PopulateMetadataIfTargetIsIEndpointMetadataProvider<TResult2>(method, builder);
}
```

Reply:
These won't be called from the source generator is my point, I guess.

Reply:
How would source generators get this metadata without calling `IEndpointMetadataProvider.PopulateMetadata`? All it will see is a return type that implements the interface. It doesn't know about these nested results. Unless there is a special case added for Results of T. But that seems unnecessary when this PR makes it AOT safe.

Reply:
The source generators can just invoke the methods statically (that's what the prototype does already).

https://github.com/davidfowl/uController/blob/7c1e8f3db1c6a20738eff792386696ff89d4d7e7/GeneratedOutput/RouteBuilderExtensions.g.cs#L136-L137

This is what we're going to do.

Reply:
Ok, so there will be a special case added in the generator for `Results<...>` then?

What is the actionable feedback here? Do you want me to revert the change to make `Results<...>` types unsafe? The change required to make it safe was just some DAM attributes. I'd prefer to leave it as it is.

If you do want it reverted, then I think suppressions will need to be added to the explicit implementations of `Results<...>.PopulateMetadata`. The static interface is explicitly implemented on these types. That means if someone does call it that there will be no feedback `PopulateMetadata` on these results is trimming/AOT unsafe.

Reply:
Mark `ResultsOfTHelper` as `RequiresUnreferencedCode` and mark `Results<...>.PopulateMetadata` as `RequiresUnreferencedCode`

Reply:
I placed `RequiresUnreferencedCode` and `RequiresGeneratedCode` on `Results<...>.PopulateMetadata` and get this error:

> Trim analysis error IL2046: Microsoft.AspNetCore.Http.HttpResults.Results<TResult1,TResult2>.Microsoft.AspNetCore.Http.Metadata.IEndpointMetadataProvider.PopulateMetadata(MethodInfo, EndpointBuilder): Member 'Microsoft.AspNetCore.Http.HttpResults.Results<TResult1,TResult2>.Microsoft.AspNetCore.Http.Metadata.IEndpointMetadataProvider.PopulateMetadata(MethodInfo, EndpointBuilder)' with 'RequiresUnreferencedCodeAttribute' implements interface member 'Microsoft.AspNetCore.Http.Metadata.IEndpointMetadataProvider.PopulateMetadata(MethodInfo, EndpointBuilder)' without 'RequiresUnreferencedCodeAttribute'. 'RequiresUnreferencedCodeAttribute' annotations must match across all interface implementations or overrides.

This error prevents the scenario I mentioned above: an explicit implementation of an interface method with a warning is never seen, as it's always called through the interface.

There are two options:
* Suppressing the warnings in `Results<...>.PopulateMetadata`. If someone calls them with AOT or trimming then they could get runtime errors
* Keep the work already done to make it trim-safe. TBH I'm not sure why it should be reverted. I still would like to keep it as is.

Reply:
I guess I don't understand the scenario where this needs to be safe.

Reply:
It's public API. We might not call it, but someone else who writes a web API framework that uses results could.

Reply:
> There are two options:

> Suppressing the warnings in Results<...>.PopulateMetadata. If someone calls them with AOT or trimming then they could get runtime errors

That isn't really a viable option. We are striving to ensure that the app behaves the same with or without trimming/AOT, or you get a warning.